nestkernel/conn_builder.cpp

```diff
@@ -74,75 +71,145 @@ nest::ConnBuilder::ConnBuilder( NodeCollectionPTR sources,
+        : ConnParameter::create( ( *syn_defaults )[ names::weight ], kernel().vp_manager.get_num_threads() );
+      register_parameters_requiring_skipping_( *weight_[ indx ] );
+      delay_[ indx ] = syn_params->known( names::delay )
+        ? ConnParameter::create( ( *syn_params )[ names::delay ], kernel().vp_manager.get_num_threads() )
+        : ConnParameter::create( ( *syn_defaults )[ names::delay ], kernel().vp_manager.get_num_threads() );
+    }
+    else if ( default_weight_[ indx ] )
+    {
+      delay_[ indx ] = syn_params->known( names::delay )
+        ? ConnParameter::create( ( *syn_params )[ names::delay ], kernel().vp_manager.get_num_threads() )
+        : ConnParameter::create( ( *syn_defaults )[ names::delay ], kernel().vp_manager.get_num_threads() );
+    }
+    register_parameters_requiring_skipping_( *delay_[ indx ] );
+
+    for ( Dictionary::const_iterator default_it = syn_defaults->begin(); default_it != syn_defaults->end();
+          ++default_it )
+    {
+      const Name param_name = default_it->first;
+      if ( skip_set.find( param_name ) != skip_set.end() )
+      {
+        continue; // weight, delay or not-settable parameter
+      }
+
+      if ( syn_params->known( param_name ) )
+      {
+        synapse_params_[ indx ][ param_name ] =
+          ConnParameter::create( ( *syn_params )[ param_name ], kernel().vp_manager.get_num_threads() );
+        register_parameters_requiring_skipping_( *synapse_params_[ indx ][ param_name ] );
+      }
+    }
+
+    // Now create dictionary with dummy values that we will use
+    // to pass settings to the synapses created. We create it here
+    // once to avoid re-creating the object over and over again.
+    if ( synapse_params_[ indx ].size() > 0 )
+    {
+      for ( thread tid = 0; tid < kernel().vp_manager.get_num_threads(); ++tid )
+      {
+
+        param_dicts_[ indx ].push_back( new Dictionary() );
+
+        for ( auto param : synapse_params_[ indx ] )
+        {
+          if ( param.first == names::receptor_type or param.first == names::music_channel
+            or param.first == names::synapse_label )
+          {
+            ( *param_dicts_[ indx ][ tid ] )[ param.first ] = Token( new IntegerDatum( 0 ) );
+          }
+          else
+          {
+            ( *param_dicts_[ indx ][ tid ] )[ param.first ] = Token( new DoubleDatum( 0.0 ) );
+          }
+        }
+      }
+    }
+    ++indx;
   }
-  register_parameters_requiring_skipping_( *delay_ );
 
   // Structural plasticity parameters
   // Check if both pre and post synaptic element are provided
-  if ( syn_spec->known( names::pre_synaptic_element ) and syn_spec->known( names::post_synaptic_element ) )
+  // Currently only possible with structural plasticity with single element syn_spec
+  if ( syn_spec[ 0 ]->known( names::pre_synaptic_element ) and syn_spec[ 0 ]->known( names::post_synaptic_element ) )
```

Comment:
I think the test here will not work in all cases. If a user places a `pre/post_synaptic_element` somewhere in `syn_spec[1:]` (borrowing Python slicing notation), we will not catch this here. I think we need to check if any of the two SP-related parameters appears in *any* `syn_spec` element. Maybe the easiest way to do this would be to use sets something like (very pseudocode):
```
all_keys = set()
for ( auto sp : syn_spec )
   all_keys = union(all_keys, sp.keys())
have_sp_keys = intersection(all_keys, set(pre_syn_elem, post_syn_elem))
if have_sp_keys and syn_spec.size() > 1:
  error
```